hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs

```diff
@@ -586,44 +675,83 @@ data IdeError = IdeError
-    deriving (Eq, Ord, Enum, Bounded, Show, Read)
-
-data BiosOptions = BiosOptions {
-    boGhcUserOptions :: [String]
-  , boLogging        :: BiosLogLevel
-  } deriving Show
-
-defaultOptions :: BiosOptions
-defaultOptions = BiosOptions {
-    boGhcUserOptions = []
-  , boLogging        = BlWarning
-  }
+instance ExceptionMonad m => ExceptionMonad (ReaderT e m) where
+  gcatch (ReaderT m) c = ReaderT $ \r -> m r `gcatch` \e -> runReaderT (c e) r
+  gmask a = ReaderT $ \e -> gmask $ \u -> runReaderT (a $ q u) e
+    where q :: (m a -> m a) -> ReaderT e m a -> ReaderT e m a
+          q u (ReaderT b) = ReaderT (u . b)
 
-fmBiosLog :: BiosLogLevel -> GM.GmLogLevel
-fmBiosLog bl = case bl of
-  BlError   -> GM.GmError
-  BlWarning -> GM.GmWarning
-  BlInfo    -> GM.GmInfo
-  BlDebug   -> GM.GmDebug
-  BlVomit   -> GM.GmVomit
+instance MonadTrans GhcT where
+  lift m = liftGhcT m
 
--- ---------------------------------------------------------------------
 
--- | Apply BiosOptions to default ghc-mod Options
-mkGhcModOptions :: BiosOptions -> GM.Options
-mkGhcModOptions bo = GM.defaultOptions
-  {
-    GM.optGhcUserOptions = boGhcUserOptions bo
-  , GM.optOutput = (GM.optOutput GM.defaultOptions) { GM.ooptLogLevel = fmBiosLog (boLogging bo) }
-  }
+instance MonadUnliftIO Ghc where
+    {-# INLINE askUnliftIO #-}
+    askUnliftIO = Ghc $ \s ->
+                  withUnliftIO $ \u ->
+                  return (UnliftIO (unliftIO u . flip unGhc s))
 
--- ---------------------------------------------------------------------
+    {-# INLINE withRunInIO #-}
+    withRunInIO inner =
+      Ghc $ \s ->
+      withRunInIO $ \run ->
+      inner (run . flip unGhc s)
+
+instance MonadUnliftIO (GhcT IdeM) where
+    {-# INLINE askUnliftIO #-}
+    askUnliftIO = GhcT $ \s ->
+                  withUnliftIO $ \u ->
+                  return (UnliftIO (unliftIO u . flip unGhcT s))
+
+    {-# INLINE withRunInIO #-}
+    withRunInIO inner =
+      GhcT $ \s ->
+      withRunInIO $ \run ->
+      inner (run . flip unGhcT s)
+
+instance MonadTransControl GhcT where
+    type StT GhcT a = a
+
+    {-# INLINABLE liftWith #-}
+    liftWith f = GhcT $ \s -> f $ \t -> unGhcT t s
+
+    {-# INLINABLE restoreT #-}
+    restoreT = GhcT . const
+
+instance MonadBaseControl IO (GhcT IdeM) where
+    type StM (GhcT IdeM) a = ComposeSt GhcT IdeM a;
+
+    {-# INLINABLE liftBaseWith #-}
+    liftBaseWith = defaultLiftBaseWith
+
+    {-# INLINABLE restoreM #-}
+    restoreM = defaultRestoreM
+
+instance MonadBase IO (GhcT IdeM) where
+
+    {-# INLINABLE liftBase #-}
+    liftBase = liftBaseDefault
+
+
+instance MonadPlus (GhcT IdeM) where
+    {-# INLINE mzero #-}
+    mzero = lift mzero
+
+    {-# INLINE mplus #-}
+    m `mplus` n = GhcT $ \s -> unGhcT m s `mplus` unGhcT n s
+
+instance Alternative (GhcT IdeM) where
+    {-# INLINE empty #-}
+    empty = lift empty
+
+    {-# INLINE (<|>) #-}
+    m <|> n = GhcT $ \s ->  unGhcT m s <|> unGhcT n s
+
+-- ghc-8.6 required
```

Comment:
Leave these here as they are way easier to understand.

Reply:
Why not use ifdefs to actually use these for 8.6+?